src/passes/GlobalStructInference.cpp

```diff
@@ -137,12 +137,9 @@ struct GlobalStructInference : public Pass {
 
       auto type = global->init->type.getHeapType();
 
-      // The global's declared type must match the init's type. If not, say if
-      // we had a global declared as type |any| but that contains (ref $A), then
-      // that is not something we can optimize, as ref.eq on a global.get of
-      // that global will not validate. (This should not be a problem after
-      // GlobalSubtyping runs, which will specialize the type of the global.)
-      if (global->type != global->init->type) {
+      // The global's declared type must be equality comparable.
+      if (auto eq = wasm::HeapTypes::eq.getBasic(type.getShared());
+          !Type::isSubType(global->type, Type(eq, Nullable))) {
```

Comment:
Not related to this PR, but I wonder if we can shorten these two lines for some kind of quick pattern "check if subtype of $TYPE, ignoring nullability and sharedness"...

Reply:
Maybe it would make sense to have `HeapType::isMaybeSharedSubType(a, b)` or even just `isEq()` for checking ability to be compared for equality. I'll keep my eye out for other places these would be useful.